tensorflow/compiler/mlir/xla/transforms/hlo_outline_fusion.cc

```diff
@@ -0,0 +1,555 @@
+    cluster_storage_.emplace_back(
+        new Cluster(cycles_graph_node_id, this));
+    return cluster_storage_.back().get();
+  }
+
+  void BuildNodeMap() {
+    for (auto& op : op_list_) {
+      auto node_id = cycle_detector_.NewNode();
+      op_to_node_id_[op] = node_id;
+      cluster_for_node_.emplace_back();
+      cluster_for_node_[node_id].Get() = MakeCluster(node_id);
+      for (int idx = 0; idx < op->getNumOperands() ; ++idx) {
+        auto operand_op = op->getOperand(idx).getDefiningOp();
+        if (operand_op == nullptr) {
+          // skip block argument
+          continue;
+        }
+        auto iter = op_to_node_id_.find(operand_op);
+        assert(iter != op_to_node_id_.end());
+        cycle_detector_.InsertEdge(iter->second, node_id);
+      }
+    }
+  }
+
+  Cluster* GetClusterForNode(Operation* n) {
+    auto id = op_to_node_id_.at(n);
+    return cluster_for_node_[id].Get();
+  }
+
+  Cluster* GetClusterForCyclesGraphNode(int node_id) {
+    if (node_id >= cluster_for_node_.size() || op_list_[node_id] == nullptr) {
+      return nullptr;
+    }
+    Cluster* cluster = cluster_for_node_[node_id].Get();
+    if (cluster) {
+      assert(cluster->cycles_graph_node_id() == node_id);
+    }
+    return cluster;
+  }
+
+  // Merge the clusters `cluster_from` and `cluster_to`.  After this step the
+  // larger combined cluster is represented by `cluster_from`'s ID in
+  // `cycles_graph_`.
+  bool MergeClusters(Cluster* cluster_from, Cluster* cluster_to) {
+    int from = cluster_from->cycles_graph_node_id();
+    int to = cluster_to->cycles_graph_node_id();
+
+    if (!cycle_detector_.ContractEdge(from, to)) {
+      llvm::dbgs() << "Could not contract " << from << " -> " << to
+                   << " because contracting the edge would create a cycle.";
+      return false;
+    }
+
+    // Merge the clusters.
+    cluster_from->Merge(cluster_to);
+
+    // Merge the UnionFind<Cluster*>.
+    cluster_for_node_[from].Merge(&cluster_for_node_[to]);
+
+    return true;
+  }
+
+  template <typename FnTy>
+  bool ForEachEdgeInPostOrder(FnTy fn) {
+    bool changed = false;
+    for (int32 node : cycle_detector_.AllNodesInPostOrder()) {
+      Cluster* cluster_from = GetClusterForCyclesGraphNode(node);
+      if (!cluster_from) {
+        continue;
+      }
+
+      // Make a copy of the set of successors because we may modify the graph in
+      // TryToContractEdge.
+      std::vector<int32> successors_copy =
+          cycle_detector_.SuccessorsCopy(cluster_from->cycles_graph_node_id());
+
+      for (int to : successors_copy) {
+        Cluster* cluster_to = GetClusterForCyclesGraphNode(to);
+        if (!cluster_to) {
+          continue;
+        }
+
+        bool contracted_edge = fn(cluster_from, cluster_to);
+        changed |= contracted_edge;
+      }
+    }
+
+    return changed;
+  }
+
+  std::vector<Value> GetResultsOfFusedPattern(Cluster* from, Cluster* to) {
+    auto fused_pattern = MergeFusionPattern(
```

Comment:
It seems wasteful to materialize these in vectors just to be able to iterate over them.